src/stdune/dyn0.ml

```diff
@@ -19,96 +22,147 @@ type t =
-  | Sexp s -> pp_sexp s
-  | Option None -> pp (Variant ("None", []))
-  | Option (Some x) -> pp (Variant ("Some", [x]))
-  | List x ->
-    Pp.box
-      (Pp.concat
-         [ Pp.char '['
-         ; Pp.concat_map ~sep:(Pp.seq (Pp.char ';') Pp.space) x ~f:pp
-         ; Pp.char ']'
-         ])
-  | Array a ->
-    Pp.box
-      (Pp.concat
-         [ Pp.verbatim "[|"
-         ; Pp.concat_map ~sep:(Pp.seq (Pp.char ';') Pp.space) (Array.to_list a) ~f:pp
-         ; Pp.verbatim "|]"
-         ])
-  | Set xs ->
-    Pp.box
-      (Pp.concat
-         [ Pp.verbatim "set {"
-         ; Pp.concat_map ~sep:(Pp.seq (Pp.char ';') Pp.space) xs ~f:pp
-         ; Pp.verbatim "}"
-         ])
-  | Map xs ->
-    Pp.box
-      (Pp.concat
-         [ Pp.verbatim "map {"
-         ; Pp.concat_map ~sep:(Pp.seq (Pp.char ';') Pp.space) xs ~f:(fun (k, v) ->
-             Pp.box
-               (Pp.concat
-                  [ pp k
-                  ; Pp.space
-                  ; Pp.verbatim ":"
-                  ; Pp.space
-                  ; pp v
-                  ])
-           )
-         ; Pp.verbatim "}"
-         ])
-  | Tuple x ->
-    Pp.box
-      (Pp.concat
-         [ Pp.char '('
-         ; Pp.concat_map ~sep:(Pp.seq (Pp.char ',') Pp.space) x ~f:pp
-         ; Pp.char ')'
-         ])
-  | Record fields ->
-    Pp.vbox ~indent:2
-      (Pp.concat
-         [ Pp.char '{'
-         ; Pp.concat_map ~sep:(Pp.char ';') fields ~f:(fun (f, v) ->
-             Pp.concat
-               [ Pp.verbatim f
-               ; Pp.space
-               ; Pp.char '='
-               ; Pp.space
-               ; Pp.box ~indent:2 (pp v)
-               ]
-           )
-         ; Pp.char '}'
-         ])
-  | Variant (v, []) -> Pp.verbatim v
-  | Variant (v, xs) ->
-    Pp.hvbox ~indent:2
-      (Pp.concat
-         [ Pp.verbatim v
-         ; Pp.space
-         ; Pp.concat_map ~sep:(Pp.char ',') xs ~f:pp
-         ])
+  let unsnoc l = match List.rev l with
+    | last :: before_last -> Some (List.rev before_last, last)
+    | [] -> None
+
+  let split s ~on =
+    let open String in
+    let rec loop i j =
+      if j = length s then
+        [sub s ~pos:i ~len:(j - i)]
+      else if s.[j] = on then
+        sub s ~pos:i ~len:(j - i) :: loop (j + 1) (j + 1)
+      else
+        loop i (j + 1)
+    in
+    loop 0 0
+
+  let string_in_ocaml_syntax str =
+    let is_space = function
+      | ' ' ->
+        (* don't need to handle tabs because those are already escaped *)
+        true
+      | _ -> false
+    in
+    let escape_protect_first_space s =
+      let first_char =
+        if String.length s > 0 && is_space s.[0]
+        then "\\"
+        else
+          " "
+      in
+      first_char ^  String.escaped s
+    in
+    (* CR-someday aalekseyev: should use the method from
+       [Dune_lang.prepare_formatter] so that the formatter can fit multiple
+       lines on one line. *)
+    match split ~on:'\n' str with
```

Comment:
Or you can even use `String_split.split_lines`

Reply:
I suspect `split_lines` ignores the trailing newline? (these things usually do)
We would have to special-case that, I think, so I'd just leave this part as-is.